X86/X86RaisedValueTracker.cpp

```diff
@@ -1179,6 +1179,17 @@ X86RaisedValueTracker::reinterpretSSERegValue(Value *SrcVal, Type *DstTy,
 
 Type *X86RaisedValueTracker::getSSEInstructionType(const MachineInstr &MI,
                                                    LLVMContext &Ctx) {
+  // TODO: temp workaround, fix this
+  switch (MI.getOpcode()) {
+  case X86::MOVDQAmr:
+  case X86::MOVDQArm:
+  case X86::MOVDQArr:
+  case X86::MOVDQUmr:
+  case X86::MOVDQUrm:
+  case X86::MOVDQUrr:
+    return VectorType::get(Type::getInt32Ty(Ctx), 4, false);
+  }
+
```

Comment:
I am not sure how to handle this. I've added this temporary workaround which in my opinion should not be merged into master. `(TSFlags & llvm::X86II::OpPrefixMask) == llvm::X86II::X` evaluates to true for the `movdq` instructions, which would return `float` instead of `<4 x i32>`, which I assume would be the correct type.

Reply:
@martin-fink Sorry for the delay. Thanks for the PRs. 

`X86MachineInstructionRaiser::getPhysRegOperandType(const MachineInstr &MI, unsigned int OpIndex)` returns the type of register operand at index `OpIndex` of `MI`. In case it is an SSE register, the call made is to `X86RaisedValueTracker::getSSEInstructionType(const MachineInstr &MI, LLVMContext &Ctx)` which does not consider the index of the operand whose type is to be constructed. There appears to be a anomaly between the way SSE register operand types are constructed and calling function's intent. That might be the reason for the need of such a work around.

I'll take a closer look at this in the next couple of days.